### ISSUE TITLE
Extracting Operator or CUDA Kernel for TensorRT/ONNX Custom Op Registration

### ISSUE THREAD

**Issue by areddy2022** [microsoft/BitBLAS#123]
Hello,

I was curious whether there is a way to register generated GPU kernels with TensorRT or ONNX to allow for the model to be optimized and deployed with the BitBLAS kernel included.

**Comment by LeiWang1999** [comment]
Hi @areddy2022 , It's difficult to integrate kernel into trt or onnxruntime, but if you want to use our end2end version (which take onnx file as inputs and do compilation optimizations), you can check out the branch https://github.com/microsoft/BitBLAS/tree/osdi24_ladder_artifact

**Comment by areddy2022** [comment]
@LeiWang1999,
Thank you, we'll checkout ladder and see if we can use the kernels it generates on Jetson.

In particular, we were curious whether it was possible to extract the generated optimized CUDA code from TVM. For example, with some TVM code for a vector add:
`nt = 64  # number of threads in a block

def parallel(n):
    A, B, C = d2ltvm.vector_add(n)
    s = te.create_schedule(C.op)
    bx, tx = s[C].split(C.op.axis[0], factor=nt)
    s[C].bind(bx, te.thread_axis("blockIdx.x"))
    s[C].bind(tx, te.thread_axis("threadIdx.x"))
    return s, (A, B, C)

s, args = parallel(256)
tvm.lower(s, args, simple_mode=True)`

We may then check the generate C code by running:
`mod = tvm.build(s, args, 'cuda')
dev_mod = mod.imported_modules[0]
print(dev_mod.get_source())`

Which shall produce:
`extern "C" __global__ void default_function_kernel0(void** __restrict__ c, void** __restrict__ a, void** __restrict__ b) {
  (( float*)c)[(((((int)blockIdx.x) * 64) + ((int)threadIdx.x)))] = ((( float*)a)[(((((int)blockIdx.x) * 64) + ((int)threadIdx.x)))] + (( float*)b)[(((((int)blockIdx.x) * 64) + ((int)threadIdx.x)))]);
}`

It is then possible to generate the PyTorch, ONNX, and TensorRT bindings (perhaps programmatically). If you think that this is possible (or useful), I am willing to try to set up a PR to add this functionality

**Comment by LeiWang1999** [comment]
absolutely we can, for example, 

```python
import bitblas
import torch

# enabling debug output

bitblas.set_log_level("Debug")
matmul_config = bitblas.MatmulConfig(
    M=1,  # M dimension
    N=1024,  # N dimension
    K=1024,  # K dimension
    A_dtype="float16",  # activation A dtype
    W_dtype="int4",  # weight W dtype
    accum_dtype="float16",  # accumulation dtype
    out_dtype="float16",  # output dtype
    layout="nt",  # matrix layout, "nt" indicates the layout of A is non-transpose and the layout of W is transpose
    with_bias=False,  # bias
    # configs for weight only quantization
    group_size=None,  # setting for grouped quantization
    with_scaling=False,  # setting for scaling factor
    with_zeros=False,  # setting for zeros
    zeros_mode=None,  # setting for how to calculating zeros
)

matmul = bitblas.Matmul(config=matmul_config)

print(matmul.get_source())
```

Then you can extract the cuda source, and if the kernel is dynamic, you can checkout the matmul.srcpath to get a wrapped source that we used to invoke from torch.
